db/compacted_db_impl.cc

```diff
@@ -72,18 +72,15 @@ std::vector<Status> CompactedDBImpl::MultiGet(const ReadOptions& options,
     }
   }
   std::vector<Status> statuses(keys.size(), Status::NotFound());
```

Comment:
It's not specific to this PR, but since we are creating a new MultiGet() API I hope we think about the API as a whole so that we don't generate yet another bad API. Instead of always doing a malloc for return status, can we allow users to pass in one and we just fill it.

Think about it further more, we have 4 vectors, CF handle, keys, values and status, all with the same size. Should we make it 4, or should we consolidate it to just 1? I don't have an answer but I think it's something we can think about more. I hope @ot doesn't want it urgently so we can spend more time to think about a good interface.

Reply:
Yes it's not urgent and I would also prefer a better interface. If you port the `folly::Range` concept, the allocation of the result containers could be completely managed by the caller, which means all allocations could be avoided if the number of results is statically bounded.

Something like
```
void CompactedDBImpl::MultiGet(
  const ReadOptions& options,
  const Range<ColumnFamilyHandle**> cfs,
  const Range<PinnableSlice*> keys,
  const Range<PinnableSlice**> values,
  const Range<Status**> statuses);
```

Reply:
@siying I'm fine with pushing the allocation of PinnableSlice vector and status vector to the caller for the sake of performance of the new MultiGet API. It's essentially a trade-off between new and old APIs, because to reduce code duplication, we will probably have old API (taking vector<string>) call the new API (taking vector<PinnableSlice>), hence the allocation overhead is passed to the old API. I guess we can do some benchmarking after the code changes are stabilized. 
In terms of merging 4 vectors into 1, I feel it makes sense if this pattern appears in multiple places, then we can create a new class to hold the group of data: (CF handles, keys, values, status). Otherwise it feels a bit overkill to group them together just for the sake of MultiGet.